File: genai_telemetry/exporters/loki.py
"""
Grafana Loki exporter.
"""

import atexit
import base64
import json
import logging
import threading
import time
import urllib.request
from typing import Any, Dict, List, Optional

from genai_telemetry.exporters.base import BaseExporter

logger = logging.getLogger("genai_telemetry.exporters.loki")


class LokiExporter(BaseExporter):
    """Sends logs to Grafana Loki."""
    
    def __init__(
        self,
        url: str = "http://localhost:3100",
        tenant_id: str = None,
        username: str = None,
        password: str = None,
        labels: Dict[str, str] = None,
        batch_size: int = 10,
        flush_interval: float = 5.0
    ):
        """
        Initialize Loki exporter.
        
        Args:
            url: Loki server URL
            tenant_id: Multi-tenant organization ID
            username: Username for basic auth
            password: Password for basic auth
            labels: Default labels for all log streams
            batch_size: Number of logs to batch before sending
            flush_interval: Seconds between automatic flushes
        """
        self.url = url.rstrip("/") + "/loki/api/v1/push"
        self.tenant_id = tenant_id
        self.username = username
        self.password = password
        self.labels = labels or {"job": "genai-telemetry"}
        self.batch_size = batch_size
        self.flush_interval = flush_interval
        
        self._batch: List[dict] = []
        self._lock = threading.Lock()
        self._flush_thread: Optional[threading.Thread] = None
        self._running = False
        
        atexit.register(self.stop)
    
    def start(self) -> None:
        """Start the exporter and background flush thread."""
        if self._running:
            return
        self._running = True
        if self.batch_size > 1:
            self._flush_thread = threading.Thread(target=self._flush_loop, daemon=True)
            self._flush_thread.start()
    
    def stop(self) -> None:
        """Stop the exporter and flush remaining data."""
        self._running = False
        self.flush()
    
    def _flush_loop(self) -> None:
        """Background thread for periodic flushing."""
        while self._running:
            time.sleep(self.flush_interval)
            self.flush()
    
    def flush(self) -> None:
        """Flush any buffered logs."""
        with self._lock:
            if not self._batch:
                return
            batch = self._batch.copy()
            self._batch = []
        self._send_batch(batch)
    
    def _get_headers(self) -> dict:
        """Build authentication headers."""
        headers = {"Content-Type": "application/json"}
        if self.tenant_id:
            headers["X-Scope-OrgID"] = self.tenant_id
        if self.username and self.password:
            credentials = base64.b64encode(f"{self.username}:{self.password}".encode()).decode()
            headers["Authorization"] = f"Basic {credentials}"
        return headers
    
    def _send_batch(self, batch: List[dict]) -> bool:
        """Send a batch of logs to Loki."""
        if not batch:
            return True
        
        # Group by labels
        streams = {}
        for span in batch:
            # Build labels for this span
            labels = {
                **self.labels,
                "span_type": span.get("span_type", "UNKNOWN"),
                "model_name": span.get("model_name", "unknown"),
                "workflow_name": span.get("workflow_name", "unknown"),
            }
            label_str = ",".join(f'{k}="{v}"' for k, v in sorted(labels.items()))
            
            if label_str not in streams:
                streams[label_str] = {"stream": labels, "values": []}
            
            # Timestamp in nanoseconds
            ts_ns = str(int(time.time() * 1e9))
            log_line = json.dumps(span)
            streams[label_str]["values"].append([ts_ns, log_line])
        
        payload = {"streams": list(streams.values())}
        data = json.dumps(payload).encode("utf-8")
        
        req = urllib.request.Request(
            self.url,
            data=data,
            headers=self._get_headers(),
            method="POST"
        )
        
        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                return resp.status in (200, 204)
        except Exception as e:
            logger.error(f"Loki Error: {e}")
            return False
    
    def export(self, span_data: Dict[str, Any]) -> bool:
        """Export a single span as a log entry."""
        if self.batch_size <= 1:
            return self._send_batch([span_data])
        
        with self._lock:
            self._batch.append(span_data)
            should_flush = len(self._batch) >= self.batch_size
        
        if should_flush:
            self.flush()
        return True

File: js/src/exporters/loki.ts
/**
 * Grafana Loki exporter.
 */

import type { SpanData } from '../types';
import { BaseExporter } from './base';

export interface LokiExporterOptions {
  url?: string;
  tenantId?: string;
  batchSize?: number;
  flushInterval?: number;
}

interface LokiStream {
  stream: Record<string, string>;
  values: Array<[string, string]>;
}

export class LokiExporter extends BaseExporter {
  private url: string;
  private tenantId?: string;
  private batchSize: number;
  private flushInterval: number;

  private batch: SpanData[] = [];
  private flushTimer?: ReturnType<typeof setInterval>;
  private running: boolean = false;

  constructor(options: LokiExporterOptions = {}) {
    super();
    
    let url = (options.url || 'http://localhost:3100').replace(/\/$/, '');
    if (!url.includes('/loki/api/v1/push')) {
      url += '/loki/api/v1/push';
    }
    
    this.url = url;
    this.tenantId = options.tenantId;
    this.batchSize = options.batchSize || 10;
    this.flushInterval = (options.flushInterval || 5) * 1000;
  }

  start(): void {
    if (this.running) return;
    this.running = true;

    if (this.batchSize > 1) {
      this.flushTimer = setInterval(() => this.flush(), this.flushInterval);
    }
  }

  stop(): void {
    this.running = false;
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = undefined;
    }
    this.flush();
  }

  async flush(): Promise<void> {
    if (this.batch.length === 0) return;

    const batchToSend = [...this.batch];
    this.batch = [];

    await this.sendBatch(batchToSend);
  }

  private spansToLokiStreams(spans: SpanData[]): LokiStream[] {
    // Group spans by their labels
    const streamMap = new Map<string, LokiStream>();

    for (const span of spans) {
      const labels: Record<string, string> = {
        job: 'genai-telemetry',
        span_type: span.span_type,
        workflow: span.workflow_name || 'default',
        status: span.status,
      };

      if (span.model_name) labels.model = span.model_name;
      if (span.model_provider) labels.provider = span.model_provider;

      // Create a key for grouping
      const labelKey = Object.entries(labels)
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([k, v]) => `${k}="${v}"`)
        .join(',');

      if (!streamMap.has(labelKey)) {
        streamMap.set(labelKey, {
          stream: labels,
          values: [],
        });
      }

      // Loki expects nanosecond timestamps as strings
      const timestamp = new Date(span.timestamp).getTime() * 1_000_000;
      const logLine = JSON.stringify(span);
      
      streamMap.get(labelKey)!.values.push([String(timestamp), logLine]);
    }

    return Array.from(streamMap.values());
  }

  private async sendBatch(spans: SpanData[]): Promise<boolean> {
    if (spans.length === 0) return true;

    const streams = this.spansToLokiStreams(spans);
    const payload = { streams };

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (this.tenantId) {
      headers['X-Scope-OrgID'] = this.tenantId;
    }

    try {
      const response = await fetch(this.url, {
        method: 'POST',
        headers,
        body: JSON.stringify(payload),
      });

      return response.ok || response.status === 204;
    } catch (error) {
      console.error('Loki Error:', error);
      return false;
    }
  }

  async export(spanData: SpanData): Promise<boolean> {
    if (this.batchSize <= 1) {
      return this.sendBatch([spanData]);
    }

    this.batch.push(spanData);

    if (this.batch.length >= this.batchSize) {
      await this.flush();
    }

    return true;
  }

  async healthCheck(): Promise<boolean> {
    try {
      const baseUrl = this.url.replace('/loki/api/v1/push', '');
      const response = await fetch(`${baseUrl}/ready`);
      return response.ok;
    } catch {
      return false;
    }
  }
}
